stubs/redis/redis/utils.pyi

```diff
@@ -1,8 +1,21 @@
-from typing import Any
+from typing import Any, ContextManager, Optional, Text, overload
+from typing_extensions import Literal
 
-HIREDIS_AVAILABLE: Any
+from .client import Pipeline, Redis
 
-def from_url(url, db=..., **kwargs): ...
-def pipeline(redis_obj): ...
+HIREDIS_AVAILABLE: Any
+@overload
+def from_url(url: Text, db: Optional[int] = ..., *, decode_responses: Literal[True], **kwargs) -> Redis[str]: ...
+@overload
+def from_url(url: Text, db: Optional[int] = ..., *, decode_responses: Literal[False] = ..., **kwargs) -> Redis[bytes]: ...
+@overload
+def str_if_bytes(value: bytes) -> str: ...
+@overload
+def str_if_bytes(value: object) -> object: ...
```

Comment:
We should use a type var for the second overload:

```python
_T = TypeVar("_T")
...
@overload
def str_if_bytes(value: _T) -> _T: ...

Reply:
That was my first thought as well but then I got the following error
> Overloaded function signatures 1 and 2 overlap with incompatible return types

From what I understood, correct me if I am wrong, the generic type could be `bytes` so we end-up with two signatures giving different return type when the argument type is `bytes`. Curious what your thoughts are?

Reply:
mypy is a bit overzealous in cases like this and we can disable the warning using `# type: ignore`. Overloads are processed in order, so if `value` is `bytes`, the first overload is matched. (Weirdly, this behavior is not described either in PEP 484 nor the typing docs, but that looks like an oversight. I will bring it up on typing-sig.)

Reply:
Interesting, learned something new today! It's a pity it's not documented indeed but I'll update the stubs to use a generic type then.